### PR TITLE
SONARJAVA-2714 Rule S3626: Jump statements should not be redundant

### PR DESCRIPTION
**Pull request by pynicolas**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/RedundantJumpCheck.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      return ((ReturnStatementTree) tree).expression() != null;
+    }
+    return false;
+  }
+
+  private boolean isSwitchCaseChild(Tree tree) {
+    return tree.parent().is(Tree.Kind.CASE_GROUP);
+  }
+
+  private Block nonEmptySuccessor(Block initialBlock) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantJumpCheck.java]: Code Smell Code Smell: Make "nonEmptySuccessor" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=1977&issues=AWKWXYy0-RUO4oGVXY64&open=AWKWXYy0-RUO4oGVXY64)

[ISSUE_KEY]: AWKWXYy0-RUO4oGVXY64

File: java-checks/src/main/java/org/sonar/java/checks/RedundantJumpCheck.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  private boolean isReturnWithExpression(Tree tree) {
+    if (tree.is(Tree.Kind.RETURN_STATEMENT)) {
+      return ((ReturnStatementTree) tree).expression() != null;
+    }
+    return false;
+  }
+
+  private boolean isSwitchCaseChild(Tree tree) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantJumpCheck.java]: Code Smell Code Smell: Make "isSwitchCaseChild" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=1977&issues=AWKWXYy0-RUO4oGVXY63&open=AWKWXYy0-RUO4oGVXY63)

[ISSUE_KEY]: AWKWXYy0-RUO4oGVXY63

File: java-checks/src/main/java/org/sonar/java/checks/RedundantJumpCheck.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+      successorWithoutJump = nonEmptySuccessor(successorWithoutJump);
+      Block successor = nonEmptySuccessor(block.successors().iterator().next());
+      if (successorWithoutJump.equals(successor)) {
+        reportIssue(terminator, "Remove this redundant jump.");
+      }
+    }
+  }
+
+  private boolean isReturnWithExpression(Tree tree) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantJumpCheck.java]: Code Smell Code Smell: Make "isReturnWithExpression" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=1977&issues=AWKWXYy0-RUO4oGVXY62&open=AWKWXYy0-RUO4oGVXY62)

[ISSUE_KEY]: AWKWXYy0-RUO4oGVXY62

File: java-checks/src/main/java/org/sonar/java/checks/RedundantJumpCheck.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  private void checkBlock(Block block) {
+    Block successorWithoutJump = block.successorWithoutJump();
+    Tree terminator = block.terminator();
+
+    if (terminator != null
+      && successorWithoutJump != null
+      && terminator.is(Tree.Kind.CONTINUE_STATEMENT, Tree.Kind.RETURN_STATEMENT)
+      && !isReturnWithExpression(terminator)
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/RedundantJumpCheck.java]: `isReturnWithoutExpression` to remove the double negation (here and in the method)

File: its/ruling/src/test/resources/jdk6/squid-S3626.json [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+383,
+],
+'jdk6:java/security/IdentityScope.java':[
+67,
+],
+'jdk6:java/security/Policy.java':[
+309,
+],
+'jdk6:java/text/DecimalFormat.java':[
+2551,
```
Comment (m-g-sonar) [inline its/ruling/src/test/resources/jdk6/squid-S3626.json]: This two issues looks like a FPs to me, if you remove the `continue` statement, you are going to hit the `break` statement immediately after, which exit the loop.

**Pull request merged**